src/state.rs

```diff
@@ -341,9 +370,56 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
+                    .is_none()
+                {
+                    let cached_resolve_dns = Arc::new(SingleFlight::new());
+                    let cached_resolve_dns_clone = cached_resolve_dns.clone();
+                    state.set_cached_resolve_dns_for_hostname(
+                        hostname.to_owned(),
+                        cached_resolve_dns_clone,
+                    );
+
+                    // Check if this is the first request of doing DNS task for hostname
+                    if cached_resolve_dns
+                        .should_resolved_dns
```

Comment:
I donot think we need this field.

Reply:
Hi @hzxuzhonghu, this filed has a confusion name, it may be like this: first_resolved_dns